app/src/main/kotlin/at/bitfire/davdroid/sync/Syncer.kt

```diff
@@ -62,62 +64,112 @@ abstract class Syncer {
+        beforeSync(provider)
+
+        // 1. find resource collections to be synced
+        val service = serviceRepository.getByAccountAndType(account.name, serviceType)
+        if (service != null)
+            for (remoteCollection in getSyncCollections(service.id))
+                remoteCollections[remoteCollection.url] = remoteCollection
+
+        // 2. update/delete local resources and determine new (unknown) remote collections
+        val newRemoteCollections = HashMap(remoteCollections)
+        for (localCollection in localCollections(provider)) {
+            val remoteCollection = remoteCollections[getUrl(localCollection)]
+            if (remoteCollection == null)
+                // Collection got deleted on server, delete obsolete local resource
+                delete(localCollection)
+            else {
+                // Collection exists locally, update local resource and don't add it again
+                update(localCollection, remoteCollection)
+                newRemoteCollections -= remoteCollection.url
+            }
+        }
+
+        // 3. create new local collections for newly found remote collections
+        for ((_, collection) in newRemoteCollections)
+            create(provider, collection)
+
+        // 4. sync local resources
+        for (localCollection in localSyncCollections(provider))
+            remoteCollections[getUrl(localCollection)]?.let { remoteCollection ->
+                syncCollection(provider, localCollection, remoteCollection)
+            }
+
+    }
+
+
+    /** All local collections of a specific type (calendar, address book, etc) */
+    abstract fun localCollections(provider: ContentProviderClient): List<CollectionType>
 
-        // use contacts provider for address books
-        val contentAuthority =
-            if (authority == context.getString(R.string.address_books_authority))
-                ContactsContract.AUTHORITY
-            else
-                authority
+    /** Sync enabled local collections of specific type */
+    abstract fun localSyncCollections(provider: ContentProviderClient): List<CollectionType>
 
-        val accountSettings by lazy { accountSettingsFactory.forAccount(account) }
-        val httpClient = lazy { HttpClient.Builder(context, accountSettings).build() }
+    abstract fun beforeSync(provider: ContentProviderClient)
 
-        // acquire ContentProviderClient of authority to be synced
+    abstract fun getSyncCollections(serviceId: Long): List<Collection>
+
+    abstract fun getUrl(localCollection: CollectionType): HttpUrl?
+
+    abstract fun delete(localCollection: CollectionType)
+
+    abstract fun update(localCollection: CollectionType, remoteCollection: Collection)
+
+    abstract fun create(provider: ContentProviderClient, remoteCollection: Collection)
+
+    abstract fun syncCollection(provider: ContentProviderClient, localCollection: CollectionType, remoteCollection: Collection)
+
+    fun onPerformSync() {
```

Comment:
Just had a (style) idea: Currently the difference between `onPerformSync` and `sync` is not clear (and I think it's historical from the sync framework).

To clearly identify a method that "does the thing", we could define `operator fun invoke()` instead of `onPerformSync()`. Then in `BaseSyncWorker` just `syncer()` instead of `syncer.onPerformSync()` (like [use cases](https://developer.android.com/topic/architecture/domain-layer#use-cases-kotlin)).

Reply:
oh, that's pretty neat :+1: